jmespath-core/src/main/java/io/burt/jmespath/function/CompareByFunction.java

```diff
@@ -31,14 +31,23 @@ public CompareByFunction() {
     if (elements.hasNext()) {
       T result = elements.next();
       T resultValue = expression.search(result);
-      boolean expectNumbers = expectNumbers(runtime, resultValue);
+      JmesPathType resultValueType = runtime.typeOf(resultValue);
+      boolean expectNumbers = true;
+      if (resultValueType == JmesPathType.STRING) {
+        expectNumbers = false;
+      } else if (resultValueType != JmesPathType.NUMBER) {
+        return runtime.handleArgumentTypeError(this, "number or string", resultValueType.toString());
+      }
       while (elements.hasNext()) {
         T candidate = elements.next();
         T candidateValue = expression.search(candidate);
-        checkType(runtime, candidateValue, expectNumbers);
-        if (sortsBefore(runtime.compare(candidateValue, resultValue))) {
-          result = candidate;
-          resultValue = candidateValue;
+        if (checkType(runtime, candidateValue, expectNumbers)) {
+          if (sortsBefore(runtime.compare(candidateValue, resultValue))) {
+            result = candidate;
+            resultValue = candidateValue;
+          }
+        } else {
+          return runtime.handleArgumentTypeError(this, expectNumbers ? "number" : "string", runtime.typeOf(candidateValue).toString());
```

Comment:
Without having really looked at it, it ought to be possible to do using a reduce-style interface, where these reduce to a single element, whereas the sort reduces to the list of pairs.